Source/Request.swift

```diff
@@ -368,11 +395,41 @@ open class Request {
         if let error = error { self.error = error }
 
         // Start response handlers
-        internalQueue.isSuspended = false
+        processNextResponseSerializer()
 
         eventMonitor?.requestDidFinish(self)
     }
 
+    func processNextResponseSerializer() {
+        guard let responseSerializer = nextResponseSerializer() else {
+            // There's a single hook right now for the multipart upload to delete the original file
+            internalQueue.isSuspended = false
+            return
+        }
+
+        serializationQueue.async {
+            let succeeded = responseSerializer()
+
+            if succeeded {
+                // Remove the respose serializer that succeeded from the queue
+                self.removeResponseSerializer()
+
+                // Move on to the next response serializer
+                self.processNextResponseSerializer()
+
+                // Open questions:
+                // 1) What if the next response serializer throws an error that trips retry?
+                //    - Does it make sense to retry after the first response serializer has succeeded and been removed?
+                //    - Should all response serializers be executed until an error trips retry or they all succeed?
+                // 2) If any response serializer trips retry, should they all be executed the next time around?
+                //    - Currently, a successful response serializer is popped from the queue when it succeeds
+                // 3) If you can retry from a response serializer, does it make sense to allow multiple response serializers?
+                //    - Supporting multiple greatly overcomplicates things
+                //    - It's not difficult to make a custom response serializer that makes multiple passes if necessary
+            }
```

Comment:
Answers:
1. My thought is that each serializer should be called and allowed to retry until they all complete, whether from success or failure, with no retry. Each serializer would be fired for every response, since the new response may lead to different output for a given serializer. If the first serializer fails repeatedly and can't be retried anymore, we may need to be able to store that error with the serializer until we're ready to call all of them. Generally, though, I think the behavior seen by the user should be all of the response serializers called from the same response.

2. I believe so, yes, as new responses would likely yield new response objects from the serializers.

3. While it does make things more complicated, it's once of the fundamental APIs of Alamofire. However, I don't know if anyone uses multiple serializers in production rather than just for debugging. I could imagine a few uses, mainly around logging (`EventMonitor` would be better for that now), so I'm not sure how disruptive of a change it would be.

Reply:
Okay @jshier, I just pushed up some good changes in 24898921. It updates the logic to always run all the response serializers, and resets the `Request` along with the response serialization state if any of the response serializers want to retry the `Request`.

It requires a couple new APIs around response serialization which are `appendResponseSerializer` and `responseSerializerDidComplete`. These essentially take the place of the `internalQueue` previously queueing up all the closures. I also removed the `internalQueue` and replaced it with a `cleanup` function instead. I thought that might be more applicable now given that's all that the `internalQueue` now does.

What I really love about this approach is that custom response serializers shouldn't have to deal with the fact that retry is even a thing. Everything can hide behind the `DataResponseSerializerProtocol` and use the `response(queue:responseSerializer:completion:)` API.